btf/traversal.go

```diff
@@ -59,17 +59,47 @@ func children(typ Type, yield func(child *Type) bool) bool {
 			if !yield(&v.Members[i].Type) {
 				return false
 			}
+			for _, t := range v.Members[i].Tags {
+				var tag Type = &declTag{t, v, i}
+				if !yield(&tag) {
+					return false
+				}
+			}
+		}
+		for _, t := range v.Tags {
+			var tag Type = &declTag{t, v, -1}
+			if !yield(&tag) {
+				return false
+			}
```

Comment:
I'm curious if it is worthwhile to compress the code by having `var tags []string` before the type switch, moving the iteration all the way down to `return true` below, and replacing these 6 lines (and other similar occurrences) with `tags = v.Tags`.

Reply:
Or a `func yieldTags(typ Type, yield func(child *Type) bool, int component, tags []string) bool` helper.

Reply:
If you look at the `&declTag{...}` bit, the last field is -1 or positive number depending on if its a tag on the main type or a sub-element. In the case of function, we produce both declTags for the main function and function params. So you can do:

```
var tags []*declTag

switch (...){
case *btf.Func:
  ...
  for _,t := v.Tags {
    tags = append(tags, &declTag{t, v, -1})
  }
  ...
  for i := funcParams {
    for_, t := range v. ParamTags[i] {
      tags = append(tags, &declTag{t, v, i})
     }
  }
  ...
}

for _, t := range tags {
  if ! yield(tag) {
    return false
   }
}
```

But then we are not winning much, and I personally find it harder to follow

Reply:
I had a slightly different thing in mind:
```go
var tags []string

switch (...){
case *btf.Func:
  ...
  tags := v.Tags // <-- trade 5 lines for 1
  ...
  for i := funcParams {
    for_, t := range v. ParamTags[i] {
      tags = append(tags, &declTag{t, v, i})
     }
  }
  ...
}

for _, t := range tags { // <-- iterate tags in a common block at the func end
  if !yield(&declTag{t, v, -1}) {
    return false
  }
}
```
It bothers me a little that we have the same 5 line block replicated again and again. But maybe you are right and it is easier to follow the way it is now.

Reply:
> I had a slightly different thing in mind

Right, that makes more sense. Let my try this, that would remove 5 loops, so I think that is better.